File: cli/src/safeyolo/mitm_addons/memory_monitor.py
"""
memory_monitor.py - Process memory and connection state observability

Tracks active connections and WebSocket sessions to identify WHERE memory
is going (per-domain, per-connection). All state is self-cleaning:
connections are removed on disconnect, WS sessions on close.

Process memory is read from /proc/self/status (zero stored state).
Periodic ops.memory events are written at 60s intervals for post-mortem analysis.

Does NOT inherit SecurityAddon - this is infrastructure, not a security sensor.
Follows the simpler LoopGuard/AdminShield pattern.

Always on, no option to disable.

Usage:
    mitmdump -s addons/memory_monitor.py
"""

import logging
import time
from dataclasses import dataclass

from mitmproxy import connection, http

from safeyolo.core.audit_schema import EventKind, Severity
from safeyolo.core.utils import sanitize_for_log, write_event

log = logging.getLogger("safeyolo.memory-monitor")


def _read_proc_memory() -> tuple[int, int]:
    """Read VmRSS and VmHWM from /proc/self/status. Returns (rss_kb, hwm_kb).

    VmHWM (High Water Mark) is the peak *resident* memory.
    VmPeak is peak *virtual* memory (includes mmap, shared libs) - misleading for OOM analysis.
    """
    rss_kb = hwm_kb = 0
    try:
        with open("/proc/self/status") as f:
            for line in f:
                if line.startswith("VmRSS:"):
                    rss_kb = int(line.split()[1])
                elif line.startswith("VmHWM:"):
                    hwm_kb = int(line.split()[1])
    except (OSError, ValueError) as exc:
        log.debug(f"Failed to read /proc/self/status: {type(exc).__name__}: {exc}")
    return rss_kb, hwm_kb


@dataclass
class ConnInfo:
    """Tracks a single client connection. Removed on disconnect."""

    domain: str | None  # Set on first request (unknown at connect time)
    started: float  # time.time() at connection open
    flow_count: int = 0  # Incremented on each request
    bytes_sent: int = 0  # Request body sizes
    bytes_received: int = 0  # Response body sizes (non-streaming only)


@dataclass
class WsInfo:
    """Tracks a single WebSocket session. Removed on ws_end."""

    domain: str
    started: float
    message_count: int = 0  # Just a counter, NOT the messages


class MemoryMonitor:
    """Track process memory and active connection state for OOM diagnostics."""

    name = "memory-monitor"

    def __init__(self):
        self._connections: dict[str, ConnInfo] = {}
        self._ws_sessions: dict[str, WsInfo] = {}
        self._rss_start_kb: int = 0
        self._started: float = 0.0
        self._last_event_time: float = 0.0
        self._total_flows: int = 0

    def running(self):
        """Capture baseline memory at startup."""
        self._rss_start_kb, _ = _read_proc_memory()
        self._started = time.time()
        self._last_event_time = time.time()
        write_event(
            "ops.startup",
            kind=EventKind.OPS,
            severity=Severity.LOW,
            summary=f"Memory monitor started (baseline RSS: {self._rss_start_kb // 1024} MB)",
            addon=self.name,
            details={"rss_start_mb": round(self._rss_start_kb / 1024, 1)},
        )
        log.info(f"Memory monitor active (baseline RSS: {self._rss_start_kb // 1024} MB)")

    def client_connected(self, client: connection.Client):
        """Track new client connection."""
        self._connections[client.id] = ConnInfo(
            domain=None,
            started=time.time(),
        )

    def client_disconnected(self, client: connection.Client):
        """Clean up connection tracking on disconnect."""
        info = self._connections.pop(client.id, None)
        if info and info.flow_count > 0:
            lifetime = int(time.time() - info.started)
            domain = info.domain or "(unknown)"
            write_event(
                "ops.memory.conn_closed",
                kind=EventKind.OPS,
                severity=Severity.LOW,
                summary=f"Connection closed: {sanitize_for_log(domain)} ({info.flow_count} flows, {lifetime}s)",
                host=domain if info.domain else None,
                addon=self.name,
                details={
                    "flow_count": info.flow_count,
                    "lifetime_s": lifetime,
                    "bytes_sent": info.bytes_sent,
                    "bytes_received": info.bytes_received,
                },
            )

    def request(self, flow: http.HTTPFlow):
        """Track request on connection, emit periodic memory event."""
        self._total_flows += 1

        conn_id = flow.client_conn.id
        info = self._connections.get(conn_id)
        if info:
            info.flow_count += 1
            if info.domain is None:
                info.domain = flow.request.host
            info.bytes_sent += len(flow.request.content or b"")

        # Periodic memory event (at most once per 60s)
        now = time.time()
        if now - self._last_event_time >= 60:
            self._last_event_time = now
            self._emit_periodic_event(now)

    def response(self, flow: http.HTTPFlow):
        """Track response bytes (non-streaming only)."""
        if not flow.response:
            return
        if getattr(flow.response, "stream", False):
            return

        conn_id = flow.client_conn.id
        info = self._connections.get(conn_id)
        if info:
            info.bytes_received += len(flow.response.content or b"")

    def websocket_start(self, flow: http.HTTPFlow):
        """Track WebSocket session start."""
        conn_id = flow.client_conn.id
        self._ws_sessions[conn_id] = WsInfo(
            domain=flow.request.host,
            started=time.time(),
        )

    def websocket_message(self, flow: http.HTTPFlow):
        """Count WebSocket messages (do NOT store them)."""
        conn_id = flow.client_conn.id
        info = self._ws_sessions.get(conn_id)
        if info:
            info.message_count += 1

    def websocket_end(self, flow: http.HTTPFlow):
        """Clean up WebSocket session tracking."""
        conn_id = flow.client_conn.id
        info = self._ws_sessions.pop(conn_id, None)
        if info:
            lifetime = int(time.time() - info.started)
            write_event(
                "ops.memory.ws_closed",
                kind=EventKind.OPS,
                severity=Severity.LOW,
                summary=f"WebSocket closed: {sanitize_for_log(info.domain)} ({info.message_count} msgs, {lifetime}s)",
                host=info.domain,
                addon=self.name,
                details={
                    "message_count": info.message_count,
                    "lifetime_s": lifetime,
                },
            )

    def _emit_periodic_event(self, now: float):
        """Write ops.memory event with current state snapshot."""
        rss_kb, peak_kb = _read_proc_memory()
        top_connections = sorted(
            self._connections.values(),
            key=lambda c: c.flow_count,
            reverse=True,
        )[:10]

        rss_mb = round(rss_kb / 1024, 1)
        write_event(
            "ops.memory",
            kind=EventKind.OPS,
            severity=Severity.LOW,
            summary=f"RSS {rss_mb}MB, {len(self._connections)} conns, {self._total_flows} flows",
            addon=self.name,
            details={
                "rss_mb": rss_mb,
                "rss_hwm_mb": round(peak_kb / 1024, 1),
                "rss_start_mb": round(self._rss_start_kb / 1024, 1),
                "active_connections": len(self._connections),
                "active_websockets": len(self._ws_sessions),
                "total_flows": self._total_flows,
                "top_connections": [
                    {
                        "domain": conn.domain or "(unknown)",
                        "flows": conn.flow_count,
                        "age_s": int(now - conn.started),
                    }
                    for conn in top_connections
                ],
            },
        )

    def get_stats(self) -> dict:
        """Return memory and connection stats for admin API."""
        rss_kb, peak_kb = _read_proc_memory()
        now = time.time()

        connections = []
        for info in sorted(self._connections.values(), key=lambda c: c.flow_count, reverse=True):
            connections.append({
                "domain": info.domain or "(unknown)",
                "flows": info.flow_count,
                "age_s": int(now - info.started),
                "bytes_sent": info.bytes_sent,
                "bytes_received": info.bytes_received,
            })

        ws_sessions = []
        for info in self._ws_sessions.values():
            ws_sessions.append({
                "domain": info.domain,
                "messages": info.message_count,
                "age_s": int(now - info.started),
            })

        return {
            "rss_mb": round(rss_kb / 1024, 1),
            "rss_hwm_mb": round(peak_kb / 1024, 1),
            "rss_start_mb": round(self._rss_start_kb / 1024, 1),
            "uptime_s": int(now - self._started) if self._started else 0,
            "total_flows": self._total_flows,
            "active_connections": len(self._connections),
            "connections": connections[:10],
            "active_websockets": len(self._ws_sessions),
            "websockets": ws_sessions,
        }


addons = [MemoryMonitor()]

File: proxy/src/memory_monitor.rs
//! Source memory/connection monitor component. Callers own reached lifecycle
//! hooks, decoded content, wall clocks and process sampling. No timer, policy
//! decision, transport activation or body/message retention lives here.

pub mod process;

#[cfg(test)]
mod tests;

use std::{
    fmt,
    sync::{Mutex, MutexGuard},
};

use indexmap::IndexMap;
use num_bigint::{BigInt, ToBigInt};
use serde_json::Value;
use zeroize::Zeroize;

use crate::{
    audit::{self, Event, Kind, Severity, Writer},
    circuits::{CircuitValue as C, ErrorKind as NumericError},
    http_content::ContentError,
    network_guard::sanitize,
};

/// The process sampler's completed result, including its source peak fallback.
#[derive(Clone)]
pub struct MemorySample {
    pub rss_kb: BigInt,
    pub peak_kb: BigInt,
}

/// The source sampler catches OSError/ValueError but not a missing value token.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SampleError {
    Index,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ErrorKind {
    Poisoned,
    Sample(SampleError),
    Content(ContentError),
    Numeric(NumericError),
    Audit(audit::ErrorKind),
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Error(ErrorKind);
impl Error {
    pub fn kind(self) -> ErrorKind {
        self.0
    }
}
impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("memory monitor operation failed")
    }
}
impl std::error::Error for Error {}
impl From<SampleError> for Error {
    fn from(error: SampleError) -> Self {
        Self(ErrorKind::Sample(error))
    }
}
impl From<ContentError> for Error {
    fn from(error: ContentError) -> Self {
        Self(ErrorKind::Content(error))
    }
}
impl From<crate::circuits::Error> for Error {
    fn from(error: crate::circuits::Error) -> Self {
        Self(ErrorKind::Numeric(error.kind()))
    }
}
pub type Result<T> = std::result::Result<T, Error>;

struct Connection {
    domain: Option<String>,
    started: f64,
    flows: BigInt,
    sent: BigInt,
    received: BigInt,
}
impl Drop for Connection {
    fn drop(&mut self) {
        self.domain.zeroize();
    }
}
struct WebSocket {
    domain: String,
    started: f64,
    messages: BigInt,
}
impl Drop for WebSocket {
    fn drop(&mut self) {
        self.domain.zeroize();
    }
}
#[derive(Default)]
struct State {
    connections: IndexMap<String, Connection>,
    websockets: IndexMap<String, WebSocket>,
    rss_start_kb: BigInt,
    started: f64,
    last_event_time: f64,
    total_flows: BigInt,
}
impl Drop for State {
    fn drop(&mut self) {
        for (mut key, _) in self.connections.drain(..) {
            key.zeroize();
        }
        for (mut key, _) in self.websockets.drain(..) {
            key.zeroize();
        }
    }
}

/// One owner per process, retained across ordinary runtime reloads. The short
/// synchronous hook callbacks must not re-enter this owner while its state is
/// locked. No state is rolled back when decoding, sampling or submission fails.
#[derive(Default)]
pub struct MemoryMonitor {
    state: Mutex<State>,
}

impl MemoryMonitor {
    pub fn new() -> Self {
        Self::default()
    }

    fn lock(&self) -> Result<MutexGuard<'_, State>> {
        self.state.lock().map_err(|_| Error(ErrorKind::Poisoned))
    }

    /// Sample baseline, then read the wall clock twice. Repeated running calls
    /// replace the baseline/times without clearing connections, WS or totals.
    pub fn running(
        &self,
        writer: &Writer,
        sample: impl FnOnce() -> std::result::Result<MemorySample, SampleError>,
        mut clock: impl FnMut() -> f64,
    ) -> Result<()> {
        let mut state = self.lock()?;
        state.rss_start_kb = sample()?.rss_kb;
        state.started = clock();
        state.last_event_time = clock();
        // Arithmetic shift is floor division by 1024, including negative input.
        let whole_mb = &state.rss_start_kb >> 10_usize;
        let mut event = event(
            "ops.startup",
            format!("Memory monitor started (baseline RSS: {whole_mb} MB)"),
        );
        event.details = object([("rss_start_mb", C::Float(megabytes(&state.rss_start_kb)?))]);
        submit(writer, event)
    }

    pub fn client_connected(&self, id: &str, clock: impl FnOnce() -> f64) -> Result<()> {
        let mut state = self.lock()?;
        let started = clock();
        state.connections.insert(
            id.into(),
            Connection {
                domain: None,
                started,
                flows: 0.into(),
                sent: 0.into(),
                received: 0.into(),
            },
        );
        Ok(())
    }

    pub fn client_disconnected(
        &self,
        id: &str,
        writer: &Writer,
        clock: impl FnOnce() -> f64,
    ) -> Result<()> {
        let mut state = self.lock()?;
        let Some((mut key, info)) = state.connections.shift_remove_entry(id) else {
            return Ok(());
        };
        key.zeroize();
        if info.flows <= BigInt::from(0) {
            return Ok(());
        }
        let lifetime = age(clock(), info.started)?;
        let domain = display_domain(info.domain.as_deref());
        let mut event = event(
            "ops.memory.conn_closed",
            format!(
                "Connection closed: {} ({} flows, {lifetime}s)",
                sanitize(domain),
                info.flows,
            ),
        );
        event.host = info
            .domain
            .as_ref()
            .filter(|value| !value.is_empty())
            .cloned();
        event.details = object([
            ("flow_count", info.flows.clone().into()),
            ("lifetime_s", lifetime.into()),
            ("bytes_sent", info.sent.clone().into()),
            ("bytes_received", info.received.clone().into()),
        ]);
        submit(writer, event)
    }

    /// Only a known connection reaches decoded-size evaluation. Total/count/
    /// first-domain assignment precede it; periodic time is sampled afterward.
    pub fn request(
        &self,
        id: &str,
        host: &str,
        writer: &Writer,
        decoded_size: impl FnOnce() -> std::result::Result<u64, ContentError>,
        clock: impl FnOnce() -> f64,
        sample: impl FnOnce() -> std::result::Result<MemorySample, SampleError>,
    ) -> Result<()> {
        let mut state = self.lock()?;
        state.total_flows += 1;
        if let Some(info) = state.connections.get_mut(id) {
            info.flows += 1;
            if info.domain.is_none() {
                info.domain = Some(host.into());
            }
            info.sent += decoded_size()?;
        }
        let now = clock();
        if now - state.last_event_time >= 60. {
            state.last_event_time = now;
            periodic(&state, now, sample()?, writer)?;
        }
        Ok(())
    }

    /// A missing/streamed response and an unknown connection never decode.
    pub fn response(
        &self,
        id: &str,
        present: bool,
        streamed: bool,
        decoded_size: impl FnOnce() -> std::result::Result<u64, ContentError>,
    ) -> Result<()> {
        if !present || streamed {
            return Ok(());
        }
        if let Some(info) = self.lock()?.connections.get_mut(id) {
            info.received += decoded_size()?;
        }
        Ok(())
    }

    pub fn websocket_start(&self, id: &str, host: &str, clock: impl FnOnce() -> f64) -> Result<()> {
        let mut state = self.lock()?;
        let started = clock();
        state.websockets.insert(
            id.into(),
            WebSocket {
                domain: host.into(),
                started,
                messages: 0.into(),
            },
        );
        Ok(())
    }

    /// One complete data-message hook; there is no payload or byte counter.
    pub fn websocket_message(&self, id: &str) -> Result<()> {
        if let Some(info) = self.lock()?.websockets.get_mut(id) {
            info.messages += 1;
        }
        Ok(())
    }

    pub fn websocket_end(
        &self,
        id: &str,
        writer: &Writer,
        clock: impl FnOnce() -> f64,
    ) -> Result<()> {
        let mut state = self.lock()?;
        let Some((mut key, info)) = state.websockets.shift_remove_entry(id) else {
            return Ok(());
        };
        key.zeroize();
        let lifetime = age(clock(), info.started)?;
        let mut event = event(
            "ops.memory.ws_closed",
            format!(
                "WebSocket closed: {} ({} msgs, {lifetime}s)",
                sanitize(&info.domain),
                info.messages,
            ),
        );
        event.host = Some(info.domain.clone());
        event.details = object([
            ("message_count", info.messages.clone().into()),
            ("lifetime_s", lifetime.into()),
        ]);
        submit(writer, event)
    }

    /// Sample memory then wall time. Source computes every sorted connection's
    /// age before slicing the result to ten; WS reports have no presentation cap.
    pub fn get_stats(
        &self,
        sample: impl FnOnce() -> std::result::Result<MemorySample, SampleError>,
        clock: impl FnOnce() -> f64,
    ) -> Result<C> {
        let state = self.lock()?;
        let memory = sample()?;
        let now = clock();
        let mut connections = Vec::with_capacity(state.connections.len());
        for info in sorted_connections(&state) {
            connections.push(object([
                ("domain", text(display_domain(info.domain.as_deref()))),
                ("flows", info.flows.clone().into()),
                ("age_s", age(now, info.started)?.into()),
                ("bytes_sent", info.sent.clone().into()),
                ("bytes_received", info.received.clone().into()),
            ]));
        }
        let mut websockets = Vec::with_capacity(state.websockets.len());
        for info in state.websockets.values() {
            websockets.push(object([
                ("domain", text(&info.domain)),
                ("messages", info.messages.clone().into()),
                ("age_s", age(now, info.started)?.into()),
            ]));
        }
        let rss = megabytes(&memory.rss_kb)?;
        let peak = megabytes(&memory.peak_kb)?;
        let baseline = megabytes(&state.rss_start_kb)?;
        let uptime = if state.started != 0. {
            age(now, state.started)?
        } else {
            0.into()
        };
        connections.truncate(10);
        Ok(object([
            ("rss_mb", C::Float(rss)),
            ("rss_hwm_mb", C::Float(peak)),
            ("rss_start_mb", C::Float(baseline)),
            ("uptime_s", uptime.into()),
            ("total_flows", state.total_flows.clone().into()),
            (
                "active_connections",
                BigInt::from(state.connections.len()).into(),
            ),
            ("connections", C::Array(connections)),
            (
                "active_websockets",
                BigInt::from(state.websockets.len()).into(),
            ),
            ("websockets", C::Array(websockets)),
        ]))
    }
}

fn sorted_connections(state: &State) -> Vec<&Connection> {
    let mut connections: Vec<_> = state.connections.values().collect();
    connections.sort_by(|a, b| b.flows.cmp(&a.flows));
    connections
}

fn periodic(state: &State, now: f64, memory: MemorySample, writer: &Writer) -> Result<()> {
    let top = sorted_connections(state);
    let rss = megabytes(&memory.rss_kb)?;
    let rss_text = C::Float(rss).render_json(false)?;
    let mut event = event(
        "ops.memory",
        format!(
            "RSS {rss_text}MB, {} conns, {} flows",
            state.connections.len(),
            state.total_flows,
        ),
    );
    let peak = megabytes(&memory.peak_kb)?;
    let baseline = megabytes(&state.rss_start_kb)?;
    let top = top
        .into_iter()
        .take(10)
        .map(|info| {
            Ok(object([
                ("domain", text(display_domain(info.domain.as_deref()))),
                ("flows", info.flows.clone().into()),
                ("age_s", age(now, info.started)?.into()),
            ]))
        })
        .collect::<Result<Vec<_>>>()?;
    event.details = object([
        ("rss_mb", C::Float(rss)),
        ("rss_hwm_mb", C::Float(peak)),
        ("rss_start_mb", C::Float(baseline)),
        (
            "active_connections",
            BigInt::from(state.connections.len()).into(),
        ),
        (
            "active_websockets",
            BigInt::from(state.websockets.len()).into(),
        ),
        ("total_flows", state.total_flows.clone().into()),
        ("top_connections", C::Array(top)),
    ]);
    submit(writer, event)
}

fn megabytes(kb: &BigInt) -> Result<f64> {
    let C::Float(value) = C::Integer(kb.clone()).divide(&1024.into())? else {
        unreachable!("integer true division returns float");
    };
    // Same exact-binary one-decimal conversion already used by request_logger.
    Ok(if value.is_finite() {
        format!("{value:.1}").parse().expect("formatted float")
    } else {
        value
    })
}
fn age(now: f64, started: f64) -> Result<BigInt> {
    let elapsed = now - started;
    elapsed.to_bigint().ok_or_else(|| {
        Error(ErrorKind::Numeric(if elapsed.is_nan() {
            NumericError::Value
        } else {
            NumericError::Overflow
        }))
    })
}
fn display_domain(domain: Option<&str>) -> &str {
    domain
        .filter(|value| !value.is_empty())
        .unwrap_or("(unknown)")
}
fn text(value: &str) -> C {
    C::Other(Value::String(value.into()))
}
fn object<const N: usize>(fields: [(&str, C); N]) -> C {
    C::Object(
        fields
            .into_iter()
            .map(|(key, value)| (key.into(), value))
            .collect(),
    )
}
fn event(name: &str, summary: String) -> Event {
    let mut event = Event::new(name, Kind::Ops, Severity::Low, summary);
    event.addon = Some("memory-monitor".into());
    event
}
fn submit(writer: &Writer, event: Event) -> Result<()> {
    writer
        .emit(event)
        .map(|_| ())
        .map_err(|error| Error(ErrorKind::Audit(error.kind())))
}
